### PR TITLE
fix to operate on cuda kernel with clang and libc++

### PR DESCRIPTION
**Pull request by jcooky**
We find a bug about `std::tuple` with nvcc.

In C++11, `std::tuple` constructor is constexpr in libstdc++, but is not constexpr in libc++.

https://github.com/pytorch/pytorch/blob/c36b77fcdad3d54227cf0fd51693eb57035002c0/aten/src/ATen/native/cuda/Loops.cuh#L109-L111

The lines have occurred crashes in CUDA with a message `scan failed with synchronize`. It is a error message of cuda initialization.

The purpose of this PR is fixed for loop in nvcc and libc++ by not using `std::tuple`.

### CODE REVIEW

File: aten/src/ATen/native/cuda/Loops.cuh [hunk 1]
```diff
@@ -102,20 +102,18 @@ static void launch_kernel(int64_t N, const func_t& f) {
...
-template <typename traits, typename index_t, std::size_t... I>
-C10_HOST_DEVICE typename traits::ArgsTuple
-dereference_impl(char* const C10_RESTRICT data[], const index_t strides[], int i,
-                 c10::guts::index_sequence<I...>) {
-  return std::make_tuple(
-      *(typename traits::template arg<I>::type*)
-        (data[I] + i * strides[I])...);
+template <typename traits, typename func_t, typename index_t, size_t... I>
+C10_HOST_DEVICE typename traits::result_type
+invoke_impl(const func_t &f, char *const data[], const index_t strides[], int i,
```
Comment (colesbury) [inline aten/src/ATen/native/cuda/Loops.cuh]: Was removing `C10_RESTRICT` necessary?

IIRC, the restrict improved the performance for some kernels.

**Comment by TomHeaven** [comment]
This pull sovled isssue https://github.com/pytorch/pytorch/issues/25693.

**Comment by ezyang** [comment]
Ping me if y'all don't have time to review this in the near future.

File: aten/src/ATen/native/cuda/Loops.cuh (hunk 1 above)
Comment (jcooky) [reply aten/src/ATen/native/cuda/Loops.cuh]: mistake..

I will test with `C10_RESTRICT`. And push it.

**Comment by jcooky** [comment]
Why is testing failed? I don't know. sorry.

And... Can this PR be merged? I think review process has very long time

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
Just waiting on CI

**Pull request closed without merging**